implementation/src/main/java/io/smallrye/mutiny/operators/multi/MultiConcatMapOp.java

```diff
@@ -0,0 +1,312 @@
+ *
+ * @param <I> the upstream value type / input type
+ * @param <O> the output value type / produced type
+ */
+public class MultiConcatMapOp<I, O> extends AbstractMultiOperator<I, O> {
+
+    private final Function<? super I, ? extends Publisher<? extends O>> mapper;
+
+    private final boolean postponeFailurePropagation;
+
+    public MultiConcatMapOp(Multi<? extends I> upstream,
+            Function<? super I, ? extends Publisher<? extends O>> mapper,
+            boolean postponeFailurePropagation) {
+        super(upstream);
+        this.mapper = mapper;
+        this.postponeFailurePropagation = postponeFailurePropagation;
+    }
+
+    @Override
+    public void subscribe(MultiSubscriber<? super O> subscriber) {
+        if (subscriber == null) {
+            throw new NullPointerException("The subscriber must not be `null`");
+        }
+        ConcatMapMainSubscriber<I, O> sub = new ConcatMapMainSubscriber<>(subscriber,
+                mapper,
+                postponeFailurePropagation);
+
+        upstream.subscribe(Infrastructure.onMultiSubscription(upstream, sub));
+    }
+
+    public static final class ConcatMapMainSubscriber<I, O> implements MultiSubscriber<I>, Subscription, ContextSupport {
+
+        private static final int STATE_NEW = 0; // no request yet -- send first upstream request at this state
+        private static final int STATE_READY = 1; // first upstream request done, ready to receive items
+        private static final int STATE_EMITTING = 2; // received item from the upstream, subscribed to the inner
+        private static final int STATE_OUTER_TERMINATED = 3; // outer terminated, waiting for the inner to terminate
+        private static final int STATE_TERMINATED = 4; // inner and outer terminated
+        private static final int STATE_CANCELLED = 5; // cancelled
+        final AtomicInteger state = new AtomicInteger(STATE_NEW);
+
+        final MultiSubscriber<? super O> downstream;
+        final Function<? super I, ? extends Publisher<? extends O>> mapper;
+        private final boolean delayError;
+
+        final AtomicReference<Throwable> failures = new AtomicReference<>();
+
+        volatile Subscription upstream = null;
+        private static final AtomicReferenceFieldUpdater<ConcatMapMainSubscriber, Subscription> UPSTREAM_UPDATER = AtomicReferenceFieldUpdater
+                .newUpdater(ConcatMapMainSubscriber.class, Subscription.class, "upstream");
+
+        final ConcatMapInner<O> inner;
+
+        ConcatMapMainSubscriber(
+                MultiSubscriber<? super O> downstream,
+                Function<? super I, ? extends Publisher<? extends O>> mapper,
+                boolean delayError) {
+            this.downstream = downstream;
+            this.mapper = mapper;
+            this.delayError = delayError;
+            this.inner = new ConcatMapInner<>(this);
+        }
+
+        // SUBSCRIPTION (called from downstream)
```

Comment:
in this comment useful?

Reply:
Not really, but it was useful when I skimmed 10 methods of the operator :)